cpp/src/matching/subg_match.cu

```diff
@@ -0,0 +1,159 @@
+ * @param  stream the cuda stream [in / optional]
+ */
+template <typename VertexT,
+          typename SizeT,
+          typename GValueT>
+gdf_error gdf_subgraph_matching_impl(gdf_graph *graph,
+                                     VertexT* subgraphs,
+                                     cudaStream_t stream = nullptr)
+{
+  static auto row_offsets_ = [](const gdf_graph* G){
+    return static_cast<const SizeT*>(G->adjList->offsets->data);
+  };
+
+  static auto col_indices_ = [](const gdf_graph* G){
+    return static_cast<const VertexT*>(G->adjList->indices->data);
+  };
+
+  static auto values_ = [](const gdf_graph* G){
+    return static_cast<const GValueT*>(G->adjList->edge_data->data);
+  };
+
+
+  static auto nrows_ = [](const gdf_graph* G){
+    return static_cast<SizeT>(G->adjList->offsets->size - 1);
+  };
+
+  static auto nnz_ = [](const gdf_graph* G){
+    return static_cast<SizeT>(G->adjList->indices->size);
+  };
+  
+  GDF_REQUIRE(graph != nullptr, GDF_INVALID_API_CALL);
+    
+  GDF_REQUIRE(graph->adjList != nullptr, GDF_INVALID_API_CALL);
+    
+  GDF_REQUIRE(row_offsets_(graph) != nullptr, GDF_INVALID_API_CALL);
+
+  GDF_REQUIRE(col_indices_(graph) != nullptr, GDF_INVALID_API_CALL);
+    
+  auto type_id = graph->adjList->offsets->dtype;
+  GDF_REQUIRE( type_id == GDF_INT32 || type_id == GDF_INT64, GDF_UNSUPPORTED_DTYPE);
+  
+  GDF_REQUIRE( type_id == graph->adjList->indices->dtype, GDF_UNSUPPORTED_DTYPE);
+  
+  const SizeT* p_d_row_offsets = row_offsets_(graph);
+  const VertexT* p_d_col_ind = col_indices_(graph);
+  const GValueT* p_d_values = values_(graph);
+
+  assert( p_d_values );
+
+  SizeT nnz = nnz_(graph);
+  SizeT nrows = nrows_(graph);
+
+  //return unused (for now)...
+  //
+  auto t_elapsed = sm(nrows,
+                      nnz,
+                      p_d_row_offsets,
+                      p_d_col_ind,
+                      p_d_values,
+                      1,
+                      subgraphs);
+    
+  return GDF_SUCCESS;
+}
+
+/**
+ * @brief Subgraph matching. 
+ * API for gunrock implementation.
+ *
+ * @param  graph input graph; assumed undirected [in]
+ * @param  subgraphs   Return number of matched subgraphs [out]
+ */
+gdf_error gdf_subgraph_matching(gdf_graph *graph,
+                                gdf_column* subgraphs)
+
+{
+  static auto row_offsets_t_ = [](const gdf_graph* G){
+    return G->adjList->offsets->dtype;
+  };
+
+  static auto col_indices_t_ = [](const gdf_graph* G){
+    return G->adjList->indices->dtype;
+  };
+
+  static auto values_t_ = [](const gdf_graph* G){
+    return G->adjList->edge_data->dtype;
+  };
+
+  
+  auto subg_dtype = subgraphs->dtype;
+  auto ro_dtype   = row_offsets_t_(graph);
+  auto ci_dtype   = col_indices_t_(graph);
+  auto v_dtype    = values_t_(graph);
+  
+  GDF_REQUIRE( subg_dtype == ci_dtype, GDF_INVALID_API_CALL);//currently Gunrock's API requires that graph's col indices and subgraphs must be same type;
```

Comment:
Do we think this is an unreasonable assumption?  If so are they going to fix this, or do we need to worry about working around this?

Reply:
I doubt it's reasonable. But that's what they have in their code. There are a few other issues that I addressed on their Github. Specifically, indexing type for tow offsets is `SizeT` rather than `VertexT`. But...we must comply with their API.